features/mbedtls/targets/TARGET_STM/aes_alt.c

```diff
@@ -0,0 +1,495 @@
+/*
+ *  Hardware aes collector for the STM32F4 family
+ *******************************************************************************
+ * Copyright (c) 2017, STMicroelectronics
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+#if defined(MBEDTLS_AES_ALT)
+
+#include <stdio.h>
+#include "cmsis.h"
+#include "string.h"
+#include "aes.h"
+CRYP_HandleTypeDef hcryp_aes;
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = (unsigned char*)v; while( n-- ) *p++ = 0;
+}
+
+/**
+ * \brief          Initialize AES context
+ *
+ * \param ctx      AES context to be initialized
+ */
+void mbedtls_aes_init( mbedtls_aes_context *ctx )
+{
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+}
+
+/**
+ * \brief          Clear AES context
+ *
+ * \param ctx      AES context to be cleared
+ */
+void mbedtls_aes_free( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+    /* Force the CRYP Periheral Clock Reset */
+    __HAL_RCC_CRYP_FORCE_RESET();
+
+    /* Release the CRYP Periheral Clock Reset */
+    __HAL_RCC_CRYP_RELEASE_RESET();
+
+    mbedtls_zeroize( ctx, sizeof( mbedtls_aes_context ) );
+
+}
+
+/**
+ * \brief          AES key schedule (encryption)
+ *
+ * \param ctx      AES context to be initialized
+ * \param key      encryption key
+ * \param keybits  must be 128, 192 or 256
+ *
+ * \return         0 if successful, or MBEDTLS_ERR_AES_INVALID_KEY_LENGTH
+ */
+#if defined(MBEDTLS_AES_SETKEY_ENC_ALT)
+int mbedtls_aes_setkey_enc( mbedtls_aes_context *ctx, const unsigned char *key,
+                            unsigned int keybits )
+{
+    switch( keybits )
+    {
+        case 128:
+          ctx->nr = 10;
+          memcpy(ctx->aes_enc_key, key, 16);
+          hcryp_aes.Init.KeySize = CRYP_KEYSIZE_128B;
+          break;
+        case 192:
+          ctx->nr = 12;
+          memcpy(ctx->aes_enc_key, key, 24);
+          hcryp_aes.Init.KeySize = CRYP_KEYSIZE_192B;
+          break;
+        case 256:
+          ctx->nr = 14;
+          memcpy(ctx->aes_enc_key, key, 32);
+          hcryp_aes.Init.KeySize = CRYP_KEYSIZE_256B;
+          break;
+       default : return( MBEDTLS_ERR_AES_INVALID_KEY_LENGTH );
+    }
+
+    /* Deinitializes the CRYP peripheral */
+    HAL_CRYP_DeInit(&hcryp_aes);
```

Comment:
every HAL_CRYP functions are returning a status (even TIMEOUT for the crpypto function itself), but most mbed functions don't have a return type. How would you like us to treat a TIMEOUT error for instance ?